### ISSUE TITLE
Consumer listeners wrong order / parallels execution

### ISSUE THREAD

**Issue by mhascak** [Tpessia/RabbitLight#3]
Hello, I have one question / problem. When I create consumer (exchange / queue) for TaskList synchronization like below. Sometimes sometimes there is a situation when producer fire events ( Created > Assigned ),  c# Listener consume this events in wrong order ( Assigned, Created ). I´m consuming events fired from JAVA spring boot application for tasklist synchronization. 
What helps was to change options (MinChannels and MaxChannels to =1). 

```csharp
    [Exchange("camunda.usertask")]
    public class UserTaskConsumer :  ConsumerBase
    {
        private readonly ILogger<UserTaskConsumer> _logger;
        private readonly IEventBus _eventBus;
        private readonly IAsyncDuplicateLocker _asyncDuplicateLocker;

        public UserTaskConsumer(
            ILogger<UserTaskConsumer> logger, 
            IEventBus eventBus,
            IAsyncDuplicateLocker asyncDuplicateLocker) // Any injected services
        {
            _logger = logger;
            _eventBus = eventBus ?? NullEventBus.Instance;
            _asyncDuplicateLocker = asyncDuplicateLocker;
        }

        /// <summary>
        /// BEHAVIOUR
        /// - usertask.notify.overdue
        /// - usertask.notify.completed
        /// - usertask.notify.assigned
        /// - usertask.notify.created
        /// - usertask.notify.deleted
        /// </summary>
        /// <param name="context"></param>
        [Queue("usertask.behaviour")] // routing key defaults to "*"
        public async Task HandleBehaviourMessage(MessageContext<IntegrationEvent<ReferencedTask>> context)
        {
            var message = context.MessageJson();

            using (await _asyncDuplicateLocker.LockAsync(message.EventObject.BusinessProcessKey))
            {
                switch (context.EventArgs.RoutingKey)
                {
                    case "usertask.notify.completed": await _eventBus.TriggerAsync(new UserTaskCompletedEvent(message.EventObject)); break;
                    case "usertask.notify.overdue": await _eventBus.TriggerAsync(new UserTaskOverDueEvent(message.EventObject)); break;
                    case "usertask.notify.assigned": await _eventBus.TriggerAsync(new UserTaskAssignedEvent(message.EventObject)); break;
                    case "usertask.notify.created": await _eventBus.TriggerAsync(new UserTaskCreatedEvent(message.EventObject)); break;
                    case "usertask.notify.deleted": await _eventBus.TriggerAsync(new UserTaskDeletedEvent(message.EventObject)); break;
                }
            }

            _logger.LogDebug("usertask.behaviour: " + context.MessageJson());
        }
  }
```

I also check your code and noticed of this:
https://github.com/Tpessia/RabbitLight/blob/0bb92c9dad2dad54365acea7b6c6c9856645d30f/source/Consumer/Manager/ConsumerManager.cs#L136-L146

You create X channels for one consumer based on MinChannels  options, so x channels = x listeners
I´m not expert for RabbitMQ, but I read  that channels are not thread safe in c# RabbitMqClients so they suggested create one channel per consumer.

Link: http://mikehadlow.blogspot.com/2013/09/rabbitmq-amqp-channel-best-practices.html

However I´m not sure If I misunderstood something. What would you suggest for this use-case of sync? Thanks for help!
I´m using default setting of RabbitMq from your  ReadMe.

**Comment by Tpessia** [comment]
I'm really sorry for not answering you early, unfortunatly I wasn't notified and just saw it now.

About the thread safety problem, indeed it is not guaranteed by the RabbitMQ.Client lib, and the only way to enable parallel processing (as far as I know) is through multiple channels.

Specifically about ordering, there is also this reference in the docs (https://www.rabbitmq.com/semantics.html#ordering) which explains that the only way to ensure message order is by the usage of a single channel per consumer, as you've already found out.

Another solution, although not the best, would be to force a requeue (through an Exception and the OnError callback) in case the "Assigned" event is received before the "Created" event.

Please let me know if you found another solution to this problem!

Thanks for your contribution!

**Comment by mhascak** [comment]
Thanks for response :). So if you agree, that one channel per consumer is solution for my problem, I have problem with implementation which I mention above. You create one chanell per all consumers if minChanells and maxChanells options are set to 1 (if I understand  your implementation well). So solution for this situation is update implementation of your library, with possibility to configure channells per consumer individually, meybe directly in consumer attribute?

**Comment by Tpessia** [comment]
Indeed that would be the best solution for the ordering problem!

Actually on the first versions of the library the min/max channels options where set to each consumer (like you proposed), and not application wise. The problem with that implementation is that with each new consumer class you would need a new set of channels, and the total number of channels used by the application would increase very quickly (which is bad for performance).

I agree that a specific consumer wise option will also be useful, so I'll try to create a new separate option for that! So you'll have the option to set the min/max for the application and also the max for the individual consumers.

I'll try to deploy a new version (2.2.0) today with that feature. After I do it I'll let you know.

**Comment by Tpessia** [comment]
Done, just released [v2.2.0](https://www.nuget.org/packages/RabbitLight/2.2.0)!

To allow the configuration we talked about, follow this example:

```csharp
[Queue(Queues.TestSerially, maxChannels: 1, routingKeys: RoutingKeys.TestSerially)]
public async Task TestSerially(MessageContext<TestMessage> context)
{
    var headers = context.Headers();
    var msg = context.Message();
    _logger.LogInformation($"Message received: {msg.Content}");
    await Task.Delay(500);
}
```

The key change is the maxChannels option that can be passed to the Queue Attribute.

Just as heads up, for another feature that I included in the same version, you can remove the "UseRabbitLight" statement from your Startup class. In this version the lib uses IHostedService to register your consumers.

If you find another bug or have any other suggestion, please let me know!
